documentation/driver_object.md

```diff
@@ -0,0 +1,152 @@
+
+## 2. Implementation details
+### 2.1 Object driver structure
+
+```
+struct driver {
+#ifdef DRV_MGT_DEBUG
+	const char              *name;
+#endif
+	const size_t            data_size;
+	const struct driver_ops *ops;
+};
+```
+
+The *driver* structure is composed of 3 fields:
+   * *name*: debug purpose only, specify the name of the driver.
+   * *data_size*: size in bytes of the driver's data given to all driver's operations. Value can be 0, if no data needed, otherwise data object is allocated by the `drivers_alloc_data` function. Refer to [driver data](#23-driver-data)
+   * *ops*: [driver's operations](#24-driver's-operations) reference
+
+### 2.2 Driver registration
+
+A new driver object is registered in the system through the `REGISTER_DRIVER` macro:
+
+`#define REGISTER_DRIVER(_name, _data_size, _ops)`
+
+The macro has 3 parameters:
+   * *_name*: name of the driver. This value is also used for debug purpose (field *name* of *struct driver*.
+   * *_data_size*: size in bytes of the driver's data.
+   * *_ops*: driver operations object.
+
+### 2.3 Driver data
+Driver's data can be allocated if the driver object field *data_size* is not 0. These data is by default allocated in OP-TEE Heap with the weak function `drivers_alloc_data`. This function can be re-implemented on platform basis to allocate data in another place (e.g. in Secure device internal RAM).
+
+One of the objective is to ensure that driver's data are saved (encrypted or not) before memory is lost during a low power mode. Data are saved and restored by the weak functions `drivers_save_data` and `drivers_restore_data`. Default functions do nothing.
+
+### 2.4 Driver operations
+
+```
+struct driver_ops {
+	TEE_Result (*init)     (void *drvdata);
+	TEE_Result (*pm_enter) (void *drvdata, enum drv_pwrmode mode, bool wait);
+	void       (*pm_resume)(void *drvdata, enum drv_pwrmode mode);
+};
+```
+
+The *driver_ops* structure propose 3 functions:
+   * *init*: this function is the only mandatory function to install the driver in the system. This function is called first during OP-TEE initialization stage. If the function is not defined or returns an error, the driver is removed from the list of system driver.
+   * *pm_enter*: this function is called to inform driver that system is going to enter in a power mode.
+   * *pm_resume*: this function is called to inform driver that system exited the power mode.
+
+Default Power modes are:
+
+```
+enum drv_pwrmode {
+	ON = 0,
+	IDLE,
+	OFF
+};
+```
+
+| Mode      | Description           |
+|:---------:|:----------------------|
+| **ON**    | Power up and running  |
+| **IDLE**  | Low Power mode idle   |
+| **OFF**   | Power OFF             |
+
+But it's still possible to redefine the `enum drv_pwrmode` to be platform specific.
+For that the include file *platform_pw.h* has to be present in which the `enum drv_pwrmode` is redefined. Then the platform must be built with the `CFG_PLAT_PWR` compilation flag.
```

Comment:
Instead of `enum drv_pwrmode` as argument for the pm callback, this could simply be a `SUSPEND` or `RESUME` info, without bothering about the platform state.
Will the `OFF` state be used?

Reply:
No, this is no applicable on all system that require intermediate state between ON/SUSPEND.
Regarding the PSCI handler, OFF is existing.